### ISSUE TITLE
[💡 FEATURE REQUEST]: Pass the entire HTTP request body to the PHP

### ISSUE THREAD

**Issue by rlanting-framna** [roadrunner-server/roadrunner#1264]
### No duplicates 🥲.

- [X] I have searched for a similar issue in our bug tracker and didn't find any solutions.

### What happened?

Sending a POST request with Content-Type `application/x-www-form-urlencoded` does not pass the original body to the PHP worker. 

Why do I need the original body? Slack apps can send requests to your own endpoint. To verify it comes from slack, you need the original body to calculate the signature. See https://api.slack.com/authentication/verifying-requests-from-slack#known_issues_and_gotchas

Actual:

```
// (roadrunner psr-worker)
$request->getBody()->rewind();
$body = $request->getBody()->getContents();
$response = new Psr7\Response();
$response->getBody()->write($body);
$response->withHeader('Content-Type', 'text/plain');
$psr7->respond($response);
```

```
$ curl http://localhost:8089 -X POST --header "Content-Type: application/x-www-form-urlencoded" --data-urlencode 'id=123'
{"id":"123"}%
```

Expected:

```
// Contents of file served with `php -S localhost:8089`
echo file_get_contents('php://input');
```

```
$ curl http://localhost:8089 -X POST --header "Content-Type: application/x-www-form-urlencoded" --data-urlencode 'id=123'
id=123%
```

In https://github.com/roadrunner-server/roadrunner/issues/1206 there's talk about this being expected behaviour, where all bodies are parsed by RR. In my case that is exactly what I don't want to happen. And I can also not find anything in the documentation about this.

### Version (rr --version)

rr version 2.10.7 (build time: 2022-07-14T09:02:29+0000, go1.18.3), OS: darwin, arch: arm64

**Comment by rustatian** [comment]
Hey @rlantingmove4mobile 👋🏻 
If I understand correctly, the problem is in JSON formatted body, am I right?
`id=123%` -> `{"id":"123"}%`

**Comment by rlanting-framna** [comment]
Correct, the problem is that RR transforms the incoming body into JSON while I need the original.

**Comment by rustatian** [comment]
> Correct, the problem is that RR transforms the incoming body into JSON while I need the original.

But they are precisely the same. Are there any problems transforming the RR's JSON formatted body to the needed format?

**Comment by rlanting-framna** [comment]
True, I could encode it myself after extracting it from the request. But that's not always the case.

If I send this as payload:

```
a=1&b=2&z=3&c=4
```

I get this as body:

```
{"a":"1","b":"2","c":"4","z":"3"}
```

As you can see, the order is not the same. Which means I will never be able to reconstruct the original. To get ahead of any more issues like this, sending along the original would always work.

**Comment by rustatian** [comment]
So, the problem is that args are sorted (which is fine because there is no RFC or standard on this). But in your case, if I understand correctly, Slack requires precisely the same URL-encoded params (which is strange). But, ok. If I turn the sorting off, would you be able to reconstruct the body?

**Comment by rlanting-framna** [comment]
> Slack requires precisely the same URL-encoded params (which is strange).

It's not strange if you consider that `sha1sum('a=1&b=2&z=3&c=4') != sha1sum('a=1&b=2&c=4&z=3')`, which is basically how signatures work.

> If I turn the sorting off, would you be able to reconstruct the body?

From what I can see, yes I would be able to reconstruct the body in my use case.

Perhaps also take into account https://github.com/roadrunner-server/roadrunner/issues/1206#issuecomment-1178710104 because there are now two use cases for getting the original body. And not having to reconstruct the body eliminates places were things can go wrong.

And then there might also be people/projects that currently depend on the json being sorted. That would break RR for them if they update. Even though you should not depend on the sorting of the arguments:

```
https://www.ietf.org/rfc/rfc4627.txt

An object is an unordered collection of zero or more name/value
pairs, where a name is a string and a value is a string, number,
boolean, null, object, or array.
```

That also means that RR should be able to randomise the order of the json and everything should work as-is. So for you to disable sorting to cater to my use case looks like the wrong solution. If you currently do an explicit sort and disable that, there is still no guarantee that the arguments are in the same order as they came in. The parser/encoder you use might do things for optimisation reasons (or any other reason).

Long story short, disabling sorting is a stop-gap, but the actual solution is to send the original data.

**Comment by rustatian** [comment]
> It's not strange if you consider that sha1sum('a=1&b=2&z=3&c=4') != sha1sum('a=1&b=2&c=4&z=3'), which is basically how signatures work.

It's still strange from my POV that someone lay on order when computing a sha1sum (which is broken). For sure, in this case, these are different strings.

> #1206 

This ticket is about a different problem. As I mentioned, the HTTP DELETE method has not defined behavior on the body parsing. Your case is about POST.

> https://www.ietf.org/rfc/rfc4627.txt

Unordered doesn't mean that you will have the same order after parsing. For example, the Go implementation uses an associated array, which can mix the keys on the bucket relocation. Thus, we (in Go for example) use a slice with the correctly ordered args if we need to preserve order.

**Comment by rustatian** [comment]
Sorry, misclick 😞

**Comment by rustatian** [comment]
> So for you to disable sorting to cater to my use case looks like the wrong solution. 

I'm just asking what will work for you. There is no intention of changing the HTTP plugin atm.

**Comment by rustatian** [comment]
But, I agree with you, that having some option to send a raw request body will help here. Something like:

```yaml
http:
    always_send_body: true
    # send_raw_body: true
```

**Comment by rlanting-framna** [comment]
> I'm just asking what will work for you. There is no intention of changing the HTTP plugin atm.

Alright, so if that's the case then the following would be the best way forward:

1) Create an issue detailing that the http plugin needs a toggle and why. Link this issue and [#1206](https://github.com/roadrunner-server/roadrunner/issues/1206) to it. This makes it visible instead of having it buried in a comment somewhere.
2) Schedule/prioritise/label it accordingly
3) Disable the sorting on the json so you are more likely to be able to reconstruct the original body.

That will fix my immediate issue, while keeping track of what needs to be done for the future and why.

**Comment by rustatian** [comment]
Have you tried to use the `$request->query` to obtain the raw query?

**Comment by rustatian** [comment]
@rlantingmove4mobile 
There is no need to create an additional issue. I'll rename this one since this is not a bug, but you need another feature like having the raw request body. Since we have at least 2 cases, we will plan it. `2.11.1` milestone is already full, so I'll try to fit it for the `2.11.2` (approx. three weeks from now).

**Comment by rlanting-framna** [comment]
> Have you tried to use the `$request->query` to obtain the raw query?

I have not looked at the query, only the body.

> @rlantingmove4mobile There is no need to create an additional issue. I'll rename this one since this is not a bug, but you need another feature like having the raw request body. Since we have at least 2 cases, we will plan it. `2.11.1` milestone is already full, so I'll try to fit it for the `2.11.2` (approx. three weeks from now).

We might disagree about bug vs feature, but you're going to fix it so it doesn't really matter. Three weeks also short enough that I can plan for it. Thanks!

**Comment by rustatian** [comment]
I briefly looked at the current parsers logic, and if everything is fine, I'll try to resolve this ticket in the `v2.11.1`.

EDIT: `v2.11.1` is planned for the following Thursday.

**Comment by rustatian** [comment]
Ok, this is what you'll get in the PHP worker from the post form values in raw mode:

![image](https://user-images.githubusercontent.com/8040338/185622232-0ce95561-2bd5-4819-960b-e84a58f66929.png)

It would be just a string; you'll have to parse it on your own.

RR will not touch the order of the keys.

**Comment by rlanting-framna** [comment]
That looks fine to me. Awesome!

**Comment by rustatian** [comment]
Please remember that the `raw_body` feature will be available only for the `application/x-www-form-urlencoded` request type. For the `multipart/form-data` it's recommended to use a `application/x-www-form-urlencoded` to send a values.
